### PR TITLE
Support Lint Warnings

### PR DESCRIPTION
**Pull request by philippfromme**
### Features

* rename _Errors_ panel to _Problems_ panel
* change title of linting toggle to _Toggle problem view_
* show errors and warnings in _Problems_ panel
* show errors and warnings on canvas (cf. https://github.com/camunda/linting/pull/32)

## BPMN - Problems

image

## BPMN - No Problems

image

## Forms - Problems

image

### Artifacts

* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/lint-warnings-test-new/camunda-modeler-lint-warnings-test-new-linux-x64.tar.gz
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/lint-warnings-test-new/camunda-modeler-lint-warnings-test-new-mac.dmg
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/lint-warnings-test-new/camunda-modeler-lint-warnings-test-new-mac.zip
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/lint-warnings-test-new/camunda-modeler-lint-warnings-test-new-win-ia32.zip
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/lint-warnings-test-new/camunda-modeler-lint-warnings-test-new-win-x64.zip

---

Requires https://github.com/camunda/linting/pull/32

### CODE REVIEW

**Comment by nikku** [comment]
Is it correct that we abuse Carbon "Notifications" (https://carbondesignsystem.com/patterns/notification-pattern/) here to style? These errors are not notifications and I have not ever seen a linting panel like this.

What made us change from the good enough UI?

________________________________
From: Philipp Fromme ***@***.***>
Sent: 25 October 2022 22:28
To: camunda/camunda-modeler ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [camunda/camunda-modeler] Support Lint Warnings (PR #3257)


  *   rename Errors panel to Problems panel
  *   show errors and warnings in Problems panel
  *   show errors and warnings on canvas (cf. camunda/linting#32<https://github.com/camunda/linting/pull/32>)

[image]<https://user-images.githubusercontent.com/7633572/197874044-c3be0b57-ae3c-4e9f-896e-6db5596c08c3.png>

________________________________

Requires camunda/linting#32<https://github.com/camunda/linting/pull/32>

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/camunda/camunda-modeler/pull/3257

Commit Summary

  *   f3aebb6<https://github.com/camunda/camunda-modeler/pull/3257/commits/f3aebb63fc3381f9e2ee80a70e91457ef78446cb> chore: update error and warning icons
  *   d54442b<https://github.com/camunda/camunda-modeler/pull/3257/commits/d54442b8259af9ad79a7c42353f61a83f56ded51> feat: show errors and warnings in linting status bar button
  *   9cf5d74<https://github.com/camunda/camunda-modeler/pull/3257/commits/9cf5d7465bf4cc850b4e2638f2e3b68e99d469c0> feat: show errors and warnings in problems panel

File Changes

(10 files<https://github.com/camunda/camunda-modeler/pull/3257/files>)

  *   M client/resources/icons/Error.svg<https://github.com/camunda/camunda-modeler/pull/3257/files#diff-23f32208c1efa133c4d1be2ed8ddeac722f965f9b7ff0d24f746ea90880c30b5> (6)
  *   M client/resources/icons/Warning.svg<https://github.com/camunda/camunda-modeler/pull/3257/files#diff-af0664e382a9bf6ee3caa520af3a45fdb68ac3ce3b57f96cf777d4503756f74b> (8)
  *   M client/src/app/tabs/Linting.js<https://github.com/camunda/camunda-modeler/pull/3257/files#diff-4ca2a5a535bf1bb08007afad625941bfc7a6b2f1e526d373be3afd8e461d0dd5> (13)
  *   M client/src/app/tabs/Linting.less<https://github.com/camunda/camunda-modeler/pull/3257/files#diff-d6b91766968f1dce7a5e421c5f09b29ce75020002218d5cc90c537bea05b2d11> (44)
  *   M client/src/app/tabs/__tests__/LintingSpec.js<https://github.com/camunda/camunda-modeler/pull/3257/files#diff-9c1d1544f26b33c8cd821ccc5752c7f24d873a152bee257aaca7cf43cc2abd84> (29)
  *   M client/src/app/tabs/panel/Panel.less<https://github.com/camunda/camunda-modeler/pull/3257/files#diff-b1f1f8585a2c3c7f8239068729bff0c621e00cc7c86c901901691153a6ddfddf> (5)
  *   M client/src/app/tabs/panel/tabs/LintingTab.js<https://github.com/camunda/camunda-modeler/pull/3257/files#diff-390126b7ae3aa39a26b0c77db156620874bcdab7d32200d53f0d2533487cc7eb> (22)
  *   M client/src/app/tabs/panel/tabs/LintingTab.less<https://github.com/camunda/camunda-modeler/pull/3257/files#diff-5e07a3beeac3c93f32dc05b48d315de030dc397fb12c9183d3194dc4cf66db5a> (90)
  *   M client/src/app/tabs/panel/tabs/__tests__/LintingTabSpec.js<https://github.com/camunda/camunda-modeler/pull/3257/files#diff-eb72935e1987cb8c2fa6b7987323f8fe5b69b1eaf6324f5b9a293ec2ea42fdca> (41)
  *   M client/src/styles/_colors.less<https://github.com/camunda/camunda-modeler/pull/3257/files#diff-38624395ffdd591e8d0de75d0170eaf6138f87524680560abcc66df6c3a708b1> (11)

Patch Links:

  *   https://github.com/camunda/camunda-modeler/pull/3257.patch
  *   https://github.com/camunda/camunda-modeler/pull/3257.diff

**Comment by philippfromme** [comment]
Alternatively, we could remove the background and border:

electron_EZbAqbuuso

Then we would basically end up with what we had before with the only notable change being that the entire line can be clicked.

Happy holidays, @nikku. 🏝️

**Comment by YanaSegal** [comment]
Thank you Phillip for sharing this preview, It looks great! I'm fine with keeping the errors/warnings without a background color as long as this layout stays the same. 

@christian-konrad and @nithinssabu, as we have a very similar panel in the web modeler, can we adopt this look?

@nikku we don't abuse we customize carbon to our own needs. :wink:

**Comment by philippfromme** [comment]
Currently, our notifications look less like notifications than the lint errors and warnings. 🤡 So, besides making lint errors and warnings look less like notifications we should also make notifications look more like notifications, so they stand out.

image

**Comment by philippfromme** [comment]
Assuming, we'd adopt IBM Carbon for notifications I can see the problem:

image

**Comment by christian-konrad** [comment]
@philippfromme @YanaSegal recent design looks great, maybe we can still reduce the line height (in Desktop Modeler only), and then it's time for a review!

@YanaSegal we could then do same sketches for Web Modeler (but without reducing line height)

**Comment by YanaSegal** [comment]
@philippfromme Regarding the notification, I agree. Let's tackle it by opening a dedicated issue.

**Comment by nikku** [comment]
Just personal opinion, but we should not put notification styles into data tables. Just not the way the UI is meant to be composed. Or is this a practice we adopt in other apps, too?

Love https://github.com/camunda/camunda-modeler/pull/3257#issuecomment-1292241207. Much simpler, less noisy, and what I'd expect as a Nico from a lint panel.

Thanks for prototyping @philippfromme :muscle: 


----

If we want to go ahead and carbonize our notifications, that should be an easy pick, too.

**Comment by philippfromme** [comment]
> what I'd expect as a Nico

😍

**Comment by philippfromme** [comment]
I will create a separate issue for the notifications.

**Comment by nithinssabu** [comment]
Some suggestions: @philippfromme @YanaSegal @christian-konrad 

1. In my opinion, the errrors will look a bit cleaner without the borders for the cells (light grey lines in between). Discussed this with @YanaSegal and we will not have these borders when we introduce this in Web Modeler to see how it looks. This is also in line with a recent design improvement we did for the error panel in the Web Modeler. (see the attached video)
2. Should we always show ID instead of names (like we currently do). In the prototype, I see that the ID is shown when name is not present, otherwise the name is shown. In my opinion, showing the names here do not add much value as the user can identify the element by clicking on the error. IDs are less likely to overflow, names are more likely to overflow. Since the problems panel will be used by the "developer" persona, it's fine to show the IDs.

https://user-images.githubusercontent.com/5987816/198274685-0679d8b9-78df-4150-b998-9a3f3e72b163.mov

**Comment by philippfromme** [comment]
> 1. In my opinion, the errrors will look a bit cleaner without the borders for the cells (light grey lines in between). Discussed this with @YanaSegal and we will not have these borders when we introduce this in Web Modeler to see how it looks. This is also in line with a recent design improvement we did for the error panel in the Web Modeler. (see the attached video)

I would at least keep the horizontal lines (cf. https://carbondesignsystem.com/components/data-table/code/#live-demo) but we can remove the vertical lines.

> 2. Should we always show ID instead of names (like we currently do). In the prototype, I see that the ID is shown when name is not present, otherwise the name is shown. In my opinion, showing the names here do not add much value as the user can identify the element by clicking on the error. IDs are less likely to overflow, names are more likely to overflow. Since the problems panel will be used by the "developer" persona, it's fine to show the IDs.

I actually agree with you. The idea of showing the label instead of the ID sounded good in theory but as you can see it's not very useful unless you show the entire label which might take a lot of space. I would be fine with removing this feature.

@christian-konrad @YanaSegal 👀

**Comment by christian-konrad** [comment]
> t's not very useful unless you show the entire label

Why? You only need the first 10 characters to identify the problem source anyways. Let's keep showing the names. No one recognizes the IDs (which are in most cases default IDs).

> In my opinion, the errors will look a bit cleaner without the borders for the cells

I don't agree because we are using a tabular layout. We need the vertical lines, too, otherwise it may look odd when the error descriptions are left aligned but there's a margin to the element labels. 

(I would rather say we should introduce a bottom border at the problems panel header in Web Modeler, since the contrast between the header and the content is too low)

Btw. we can still change it later again if we don't see fit.

**Comment by philippfromme** [comment]
@christian-konrad Okay, so basically leave it as it is?

1. Horizontal and vertical lines for layout ✔️ 
2. Show name instead of ID if exists ✔️ 

I'm fine with that.

**Comment by YanaSegal** [comment]
@christian-konrad 

> Why? You only need the first 10 characters to identify the problem source anyways. Let's keep showing the names. No one recognizes the IDs (which are in most cases default IDs).
The names are not unique and can appear multiple times as "no" or "yes" elements. are you sure you want to show only the names?

I agree that the borders are needed

**Comment by philippfromme** [comment]
For the review, I have added a plugin that produces warnings for manual tasks.

Download here:

* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/lint-warnings-test/camunda-modeler-lint-warnings-test-linux-x64.tar.gz
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/lint-warnings-test/camunda-modeler-lint-warnings-test-mac.dmg
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/lint-warnings-test/camunda-modeler-lint-warnings-test-mac.zip
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/lint-warnings-test/camunda-modeler-lint-warnings-test-win-ia32.zip
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/lint-warnings-test/camunda-modeler-lint-warnings-test-win-x64.zip

**Comment by philippfromme** [comment]
> As discussed today in the Weekly, let's leave the IDs as are and reiterate.
> 
> @philippfromme please create a new issue for following up on this.

Follow-up issue: https://github.com/camunda/camunda-modeler/issues/3270

**Comment by philippfromme** [comment]
I created an issue for implementing this feature in the web modeler: https://github.com/camunda/web-modeler/issues/2812 @CatalinaMoisuc 👀

File: client/src/app/tabs/Linting.js [hunk 1]
```diff
@@ -28,22 +29,24 @@ export function Linting(props) {
...
         css.Linting,
         'btn',
         { 'btn--active': panel.open && panel.tab === 'linting',
-          'has-errors': errors > 0
+          'has-errors': errors > 0,
+          'has-warnings': warnings > 0
         }
       ) }
       onClick={ onToggleLinting }
       title="Toggle Errors"
```
Comment (christian-konrad) [inline client/src/app/tabs/Linting.js:46]: Tooltip should also say "Toggle problems view"

**Review by christian-konrad — CHANGES_REQUESTED** [review]
Quick change needed on the tooltip (for a11y), otherwise great!

File: client/src/app/tabs/panel/tabs/LintingTab.less [hunk 2]
```diff
@@ -1,52 +1,55 @@
+:root {
+  --linting-issue-background-color: var(--color-grey-225-10-95);
+  --linting-issue-border-color: var(--color-grey-225-10-95);
+
+  --linting-issue-padding: 8px;
```
Comment (christian-konrad) [inline client/src/app/tabs/panel/tabs/LintingTab.less:5]: `--linting-issue-padding: 6px 8x;` as agreed on before to reduce the line height

File: client/src/app/tabs/panel/tabs/LintingTab.js [hunk 3]
```diff
@@ -72,17 +72,25 @@ function LintingIssue(props) {
...
+  return <div
+    onClick={ onClick }
+    className={ classnames(css.LintingIssue, 'linting-issue', {
+      'linting-issue--error': category === 'error',
+      'linting-issue--warning': category === 'warn'
+    }) }>
+    <div className="linting-issue__header">
+      { category === 'error' ? <ErrorIcon width="16" height="16" /> : null }
+      { category === 'warn' ? <WarningIcon width="16" height="16" /> : null }
+      <span className="linting-issue__label">{ name || id }</span>
```
Comment (christian-konrad) [inline client/src/app/tabs/panel/tabs/LintingTab.js:90]: For some reason, your demo warning does not show the name, but the ID:

https://user-images.githubusercontent.com/56470362/199668465-7fe52680-8fcc-4386-81db-a9bb9823020f.mov

**Review by christian-konrad — CHANGES_REQUESTED** [review]
Found 2 more issues

**Comment by christian-konrad** [comment]
As a follow up, I also created https://github.com/camunda/camunda-modeler/issues/3272

File: client/src/app/tabs/panel/tabs/LintingTab.js (hunk 3 above)
Comment (philippfromme) [reply client/src/app/tabs/panel/tabs/LintingTab.js:90]: This is expected behavior. Unless the rule specifies `name` as part of the report the ID will be shown. I took the rule from our standard bpmnlint rules which do not have this feature yet.

File: client/src/app/tabs/Linting.js (hunk 1 above)
Comment (philippfromme) [reply client/src/app/tabs/Linting.js:46]: 👍🏻

**Comment by philippfromme** [comment]
* changed the padding to reduce line height
* changed the title of the linting toggle

image

**Comment by philippfromme** [comment]
@barmac @marstamm A technical review would be awesome! 🚀

**Review by barmac — APPROVED** [review]

**Comment by philippfromme** [comment]
Follow-up for sorting reports by category: https://github.com/camunda/camunda-modeler/pull/3275

**Pull request merged**